### PR TITLE
Add ability to patch in user-defined functionality

### PR DESCRIPTION
**Pull request by bsculac**
## Description

This PR adds a utility to patch in functionality during the execution of ARMI. Four hooks were added to the codebase to execute user-defined code from an input file. This input file has a required format that is described in the testing script for the module.

This module should allow for testing features without modifying the codebase directly, or modifying plant parameters in scoping studies from a centralized file.

---

## Checklist

- [x] The code is understandable and maintainable to people beyond the author.
- [x] Tests have been added/updated to verify that the new or changed code works.
- [x] There is no commented out code in this PR.
- [x] The commit message follows [good practices](https://terrapower.github.io/armi/developer/tooling.html).
- [x] All docstrings are still up-to-date with these changes.

If user exposed functionality was added/changed:

- [ ] Documentation added/updated in the `doc` folder.
- [ ] New or updated dependencies have been added to `setup.py`.

### CODE REVIEW

File: armi/utils/monkeyPatcher.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# Copyright 2019 TerraPower, LLC
```
Comment (keckler) [inline armi/utils/monkeyPatcher.py:1]: Since this is a new file, the copyright should be 2022.

File: armi/utils/monkeyPatcher.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    @checkPatchFlag
+    def applyPostInterfacePatch(self, upper_globals, upper_locals):
+        """
+        Attempt to apply the post-interface patch specified in the user patch file.
+        """
+        try:
+            self.userPatch.postInterfacePatch(upper_globals, upper_locals)
+        except Exception as err:
+            # TODO: Log error here
```
Comment (john-science) [inline armi/utils/monkeyPatcher.py:150]: Probably best to just do the logging, or remove these `TODO`s, yeah?

File: armi/utils/monkeyPatcher.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        self.patchFlag = True
+        if self.patchPath == "":
+            # TODO: Log that no patch file was provided.
+            self.patchFlag = False
+            return
+        if not os.path.isfile(self.patchPath):
+            self.patchFlag = False
+            raise IOError("The provided path to the patch file is invalid.")
+        self.hash = getFileSHA1Hash(self.patchPath)
+        # TODO: log the hash somewhere.
```
Comment (john-science) [inline armi/utils/monkeyPatcher.py:115]: Our goal, generally, is to not add new `TODO`s to the code.

https://terrapower.github.io/armi/developer/standards_and_practices.html#general-do-s-and-don-ts

Certainly, these all seem like easy `TODO`s to solve.

**Comment by john-science** [comment]
@bsculac 

So, are these files not saved to the database?  Just curious.

**Comment by bsculac** [comment]
@john-science The patch files specified by the user are not, should they be? The idea is that these files are specified on-the-fly by someone running ARMI, I wouldn't think it would make sense to catalog them somewhere. Unless storage in the database is something temporary (I admit ignorance on The Database and its purpose).

**Comment by john-science** [comment]
Please add a tutorial for this functionality to the Docs.

**Comment by bsculac** [comment]
Closed in favor of implementing a solution using plugins.

**Pull request closed without merging**